### DISCUSSION TITLE
Quorum queue segment consume all disc space

### DISCUSSION THREAD

**Discussion by PaniniHode** [rabbitmq/rabbitmq-server#12147] [Other] [answered]
### Describe the bug

Facing the exact same issue as in this discussions/bugs:
https://github.com/rabbitmq/rabbitmq-server/issues/6447
https://github.com/rabbitmq/discussions/issues/164
https://github.com/rabbitmq/rabbitmq-server/discussions/5258

In some random moment(Can't find the pattern) queue start to grown until disk is full(Or reach the disk watermark), one queue can consume 35-70GB from nothing, rps is low(About 100-1000 rps), total stored messages less then 10k and consumers are in reasonable amounts ~10-20. It happened about 5 times in two month.

server version is 3.13.6 official alpine based docker image.

What strange is, reduction(In runtime queue metrics) shows huge rps, always 5+ millions and it always goes down, when node is reach storage quota
<img width="662" alt="Screenshot 2024-08-28 at 1 16 50 PM" src="https://github.com/user-attachments/assets/0b803eb3-d6ee-459f-82fd-b760e0305e7d">

### Reproduction steps

As i said earlier, we can't find pattern, it's just happened and ruined one of the cluster nodes 

### Expected behavior

Normal disk usage :)

**Comment by mkuratczyk** [comment]
Would it be possible to share the queue folder with all those segment files? That should allow us to understand what special sequence of events leads to this. You can contact us directly on slack/discord to avoid making them public.

**Comment by PaniniHode** [comment]
> Would it be possible to share the queue folder with all those segment files? That should allow us to understand what special sequence of events leads to this. You can contact us directly on slack/discord to avoid making them public.

I have already deleted this queues, because node was down, but in the next time i think i can do that, but it's about 40GB, you need all of that or something specific?

**Reply by kjnilsson** [reply]
Did the data grow on only one node or on all nodes that the quorum queue spans?

**Reply by PaniniHode** [reply]
On all queue members

**Reply by kjnilsson** [reply]
If that happens again please capture the output of `rabbitmq-queues quorum_status` for the given queue.

then try to restart all consumers for the queue such that their connections are terminated and re-established.

Please also share the full configuration that is in place for the quorum queue, e.g. what delivery-limit, overflow and such settings are in place.

And if possible share segment data from one of the nodes.

**Comment by mkuratczyk** [comment]
Probably the oldest segment file will be the most interesting but please copy everything just in case we need more than that.

**Comment by PaniniHode** [comment]
> Probably the oldest segment file will be the most interesting but please copy everything just in case we need more than that.

Copy that

**Comment by michaelklishin** [comment]
We don't have any specifics to work with => this belongs to a discussion @PaniniHode

**Reply by PaniniHode** [reply]
In some reasons github don't send me any notification, sorry for the [duplicate](https://github.com/rabbitmq/rabbitmq-server/issues/12159)

**Comment by Eyald6** [comment]
We've experienced something similar, `quorum_status` shows that the `Snapshot Index` is much lower than the `Commit Index`:
```
│ Node Name          │ Raft State │ Log Index │ Commit Index │ Snapshot Index │ Term │ Machine Version │
│ rabbitmq@... │ follower   │ 90609276  │ 90609276     │ 11872694       │ 2    │ 3               │
├────────────────────┼────────────┼───────────┼──────────────┼────────────────┼──────┼─────────────────┤
│ rabbitmq@... │ follower   │ 90609276  │ 90609276     │ 11872694       │ 2    │ 3               │
├────────────────────┼────────────┼───────────┼──────────────┼────────────────┼──────┼─────────────────┤
│ rabbitmq@... │ leader     │ 90609276  │ 90609276     │ 11872694       │ 2    │ 3
```

**Reply by Eyald6** [reply]
Just to follow up here, as it might help someone.
We noticed that this issue was completely our fault! The issue was actually an edge-case in our component, we had a very old (single!) message, that we did not ack in our component, turns out every time we got it we held it without acking it. Newer messages were acked correctly which made it harder to identify the issue as the system worked properly.
We finally figured it out due to the **connection error** you get if you hold an unacknowledged message for too long. **Every 4 hours we get this error**, which means that the component simply restarts, receives this message again but holds it in (without acknowledging).
The error is pretty indicative, if the library you're using behaves nicely (we use rust's `Lapin`, which outputs `delivery acknowledgement on channel 1 timed out`).
Fixing this issue and acknowledging the old message solved this problem.

Hope this helps :)

**Reply by lukebakken** [reply]
Thank you for the follow-up. Note that this is documented - https://www.rabbitmq.com/docs/consumers#acknowledgement-timeout

**Reply by Eyald6** [reply]
Yes, we also 100% knew this, we just missed that this is actually the case here 😜

**Reply by Eyald6** [reply]
@lukebakken Would you say this is expected though? Having a single, old unacked message causing _all of the messages since_ to be persistent as well on the disk? Blindly, I would assume that only the unacked message would have to stay persistent on the disk (and not all messages _since_), though I understand it could get complicated :)

All related metrics are also not conveying the full story, for example - `queue_messages_persistent_bytes`  which I would expect to represent the space on disk the queue is taking. The only metric that does help here (but is a little too broad and indirect) is `disk_space_available_bytes`.

**Reply by michaelklishin** [reply]
@Eyald6 yes, this is expected. Quorum queues are based on a sequential log of operations. Enqueueing, "checking out" a message, acknowledgements are stored as commands in that log. 

This is why [delivery acknowledgement timeout](https://www.rabbitmq.com/docs/consumers#acknowledgement-timeout) exists: it is protection mechanisms to ensure that acknowledged messages are requeueing if not acknowledged in a certain period of time.

A new metric such as "oldest unacknowledged message" can be introduced. Ever since the delivery acknowledgement timeout was introduced, this has become a less common problem.

**Reply by Eyald6** [reply]
@michaelklishin 
I completely understand that mechanism, I'm just clarifying what I'm trying to understand here:
If I have messages 1-10 in a queue, and **I acknowledged all messages from 2-10** but **did not acknowledge message 1**, it is expected that all of the messages (1-10) are still persistent on disk.

Because if that is the expectation, I think that it could be more clear in the documentation, as well as maybe expose a metric that shows the actual persistent bytes on disk (i.e - bytes of all 1-10 messages, instead of only the unacked message)

**Reply by michaelklishin** [reply]
@Eyald6 if message 1 is not acknowledged, then the quorum queue that contains/-ed those messages will not be able to compact its Raft log and reclaim the disk space for subsequent messages, even if they have been delivered and then acknowledged.

The oldest unacknowledged message in a quorum queue will be the point up to which the log can be compacted. This is [mentioned in the Quorum Queues guide](https://www.rabbitmq.com/docs/quorum-queues#resource-use).

This generally does not apply for streams and Khepri because they have very different operational characteristics: streams have non-destructive consumers, and Khepri does not have external clients that can potentially hog a delivery and not acknowledge it.

**Reply by Eyald6** [reply]
I see, an `oldest_unacknowledged_message` would be suffice in that case, to monitor these cases (this could also happen when a message is `nacked` intentionally, in which case the acknowledgement timeout will not be triggered)

Thank you for clarifying 👑

**Reply by michaelklishin** [reply]
I have shared this suggestion with the team, we will consider it because its usefulness is fairly clear. In fact, even if QQs supported parallel (to consumer activity) compaction today.

**Reply by kjnilsson** [reply, marked as answer]
An this is why you should use a delivery limit and why we have set a default for delivery limit in 4.0.

https://www.rabbitmq.com/docs/quorum-queues#poison-message-handling

https://www.rabbitmq.com/blog/2024/08/28/quorum-queues-in-4.0#delivery-limit-is-now-20-by-default
